File: Python/ch11/PredictWinnerMemory.py
def predict_winner(arr, l, r, memory):
    if memory[l][r] is not None:
        return memory[l][r]
    if l == r:
        memory[l][r] = arr[l]
        return memory[l][r]
    a = 0
    if memory[l + 1][r] is not None:
        a = memory[l + 1][r]
    else:
        a = predict_winner(arr, l + 1, r,memory)

    b = 0
    if memory[l][r - 1] is not None:
        b = memory[l][r - 1]
    else:
        b = predict_winner(arr, l, r - 1,memory)

    return max(arr[l] - a, arr[r] - b)


if __name__ == '__main__':
    a = [1, 5, 8, 6]
    m = [[None] * (len(a) + 1) for i in range(len(a) + 1)]
    if predict_winner(a, 0, len(a) - 1, m) >= 0:
        print("玩家一获胜")
    else:
        print("玩家二获胜")

    a = [7, 13, 1, 5]
    if predict_winner(a, 0, len(a) - 1, m) >= 0:
        print("玩家一获胜")
    else:
        print("玩家二获胜")

    a = [1, 13, 1]
    if predict_winner(a, 0, len(a) - 1, m) >= 0:
        print("玩家一获胜")
    else:
        print("玩家二获胜")

File: Java/src/ch11/PredictWinnerMemory.java
package ch11;

public class PredictWinnerMemory {

    public static int predictWinner(int[] arr,int L,int R,Integer[][] memory){
        if (null != memory[L][R]){
            return memory[L][R];
        }
        if (L == R){
            memory[L][R] = arr[L];
            return memory[L][R];
        }
        int a = 0;
        if (null != memory[L+1][R]){
            a = memory[L+1][R];
        }else {
            a = predictWinner(arr,L+1,R,memory);
        }
        int b = 0;
        if (null !=  memory[L][R-1]){
            b = memory[L][R-1];
        }else {
            b = predictWinner(arr,L,R-1,memory);
        }
        memory[L][R] = Math.max(arr[L] - a,arr[R] - b);
        return memory[L][R];
    }

    public static void main(String[] args) {
        int[] arr = {1,5,8,6};
        Integer[][] memory = new Integer[arr.length][arr.length];
        if (predictWinner(arr,0,arr.length-1,memory) >= 0){
            System.out.println("玩家一获胜");
        }else{
            System.out.println("玩家二获胜");
        }


        arr = new int[]{7, 13, 1, 5};
        memory = new Integer[arr.length][arr.length];
        if (predictWinner(arr,0,arr.length-1,memory) >= 0){
            System.out.println("玩家一获胜");
        }else{
            System.out.println("玩家二获胜");
        }

        arr = new int[]{1, 13, 1};
        memory = new Integer[arr.length][arr.length];
        if (predictWinner(arr,0,arr.length-1,memory) >= 0){
            System.out.println("玩家一获胜");
        }else{
            System.out.println("玩家二获胜");
        }
    }

}
